### DISCUSSION TITLE
Node ace build extremely slow on external dev server

### DISCUSSION THREAD

**Discussion by backspacerhino** [adonisjs/core#2627] [Help]
Hi,

I've recently updated to the latest version of adonis v5 and I've noticed 2 problems when running `node ace build` or `node ace build --production`

**1. When running `node ace build` on external server it takes 2-3 minutes for it to build successfully.**

Using the same codebase it takes:
- few seconds on local computer
- few seconds on external gitlab runner (private)
- 2-3 minutes running on external server

External gitlab runner and external server are 2 different servers of the same capacity and from the same hosting company.

**2. `node ace build` does not exit with code 1 upon any ts compile errors**

If there are any compile errors, adonis will "complete" the build but will not notify the user of errors and this can and has caused pipelines to pass successfully even tho they shouldn't.

I've already notified Adonis team of this happening. 

Has anyone else experienced these issues?

Also, my package.json

```
{
  "name": "app",
  "version": "2.0.0",
  "private": true,
  "scripts": {
    "build": "node ace build",
    "start": "node ace serve --watch",
    "lint": "eslint . --ext=.ts"
  },
  "devDependencies": {
    "@adonisjs/assembler": "^5.2.0",
    "@types/node": "^14.14.35",
    "@types/socket.io-client": "^1.4.35",
    "@types/supertest": "^2.0.10",
    "@types/jsdom": "^16.2.6",
    "@types/socket.io": "^2.1.8",
    "adonis-preset-ts": "^1.0.4",
    "eslint": "^7.3.1",
    "eslint-plugin-adonis": "^1.0.13",
    "execa": "^5.0.0",
    "get-port": "^5.1.1",
    "japa": "^3.1.1",
    "pino-pretty": "^4.0.0",
    "socket.io-client": "^2.3.0",
    "supertest": "^6.1.3",
    "typescript": "^4.2.2",
    "youch": "^2.0.10",
    "youch-terminal": "^1.0.0"
  },
  "dependencies": {
    "@adonisjs/ace": "^8.0.0",
    "@adonisjs/auth": "^8.0.3",
    "@adonisjs/core": "^5.1.6",
    "@adonisjs/fold": "^7.0.9",
    "@adonisjs/lucid": "^14.0.2",
    "@adonisjs/mail": "^7.1.1",
    "@adonisjs/redis": "^5.0.9",
    "@adonisjs/repl": "^3.1.2",
    "@adonisjs/require-ts": "^1.1.1",
    "@adonisjs/session": "^6.0.3",
    "@adonisjs/shield": "^7.0.2",
    "@adonisjs/view": "^6.0.1",
    "apple-signin-auth": "^1.0.6",
    "form-data": "^3.0.0",
    "google-auth-library": "^6.0.5",
    "got": "^11.4.0",
    "jsdom": "^16.4.0",
    "mysql": "^2.18.1",
    "nyc": "^15.1.0",
    "pg": "^8.3.3",
    "phc-argon2": "^1.0.10",
    "proxy-addr": "^2.0.6",
    "reflect-metadata": "^0.1.13",
    "rethinkdb": "^2.4.2",
    "rethinkdb-ts": "^2.4.5",
    "socket.io": "^2.3.0",
    "source-map-support": "^0.5.19",
    "sqlite3": "^5.0.1",
    "ws": "^7.3.1"
  }
}
```

**Comment by thetutlage** [comment]
Can you just run `tsc` and see if that takes the same amount of time or not on the external server? This is just to distinguish if Typescript is taking time or is it something else.

```sh
./node_modules/.bin/tsc
```

Next, you can try with `noEmit` option to see if writing the files on the disk is taking time or not. Non SSD disks will be slow

```sh
./node_modules/.bin/tsc --noEmit
```

Finally, run with TypeScript diagnostics. https://www.typescriptlang.org/tsconfig/#diagnostics

```sh
./node_modules/.bin/tsc --diagnostics
```

**Reply by backspacerhino** [reply]
`./node_modules/.bin/tsc` hangs (stopped with Ctrl+C at 5m )
`./node_modules/.bin/tsc --noEmit` hangs (stopped with Ctrl+C at 5m )
`./node_modules/.bin/tsc --diagnostics` hangs (stopped with Ctrl+C at 7m )

I've tried with another adonis v5  app that is on the same server (different folder) that is using preview version of Adonis and everything behaves as it should. 


I've tried downgrading TS version to 4.0.2 (this is used by mentioned another app) but it still takes ~2m for compile

**Reply by backspacerhino** [reply]
I've upgraded my external server and tried again with every command:

`./node_modules/.bin/tsc --noEmit --diagnostics`

```
Files:             1921
Lines:           180414
Nodes:           749623
Identifiers:     259738
Symbols:         138633
Types:               76
Instantiations:       0
Memory used:    258351K
I/O read:         0.18s
I/O write:        0.00s
Parse time:       5.91s
Bind time:        1.98s
Check time:       0.00s
Emit time:        0.00s
Total time:       7.89s
```

and

`./node_modules/.bin/tsc  --diagnostics`

```
Files:             1921
Lines:           180414
Nodes:           749623
Identifiers:     259738
Symbols:         292288
Types:            63193
Instantiations:  195723
Memory used:    353113K
I/O read:         0.91s
I/O write:        0.47s
Parse time:       6.12s
Bind time:        2.18s
Check time:      10.44s
Emit time:       16.43s
Total time:      35.18s
```

Server does have an SSD.

**Reply by thetutlage** [reply]
Yup, it is TypeScript compilation that is taking time. Something you need to discuss with the TypeScript team. It is beyond my control

**Reply by backspacerhino** [reply]
Thanks for the reply. What I found weird is that with 5.0.4-preview-rc version it was a lot faster. Typescript version is the same.

I've found this link: https://stackoverflow.com/a/36737847/4555911
that might explain the difference in ts time needed between the 5.0.4 and 5.1.6 versions. 

> The second part of the initialization || (() => Observable.of(null)); was causing the slowness. Commenting it out or adding a type annotation got the compile time back down. Since Observable is generic, it seems like the TypeScript compiler needs a hint to narrow down some type checks it needs to do.

Do you have similiar usages of Observables or other Typescript Generics in Adonis core(or other updated packages) that could benefit from a more detailed type annotation?

**Reply by thetutlage** [reply]
AdonisJS uses a lot of Generics and does not use any Observables
